### PR TITLE
[Op] Completing the unary operators and refactor test

### PR DESCRIPTION
**Pull request by SiriusNEO**
This PR includes:
- Introduce `R.abs` and its legalization (For L1Loss)
- Registering most of the unary operators in [DataAPI](https://data-apis.org/array-api/draft/API_specification/elementwise_functions.html) (Without legalization)
- Split unary arith oprators and check operators (e.g. `isnan`)
- Refactor `test_tvmscript_parser`, `test_op_unary` and `test_op_binary` using `tvm.testing.parameters()`.

### CODE REVIEW

File: python/tvm/relax/op/unary.py [hunk 1]
```diff
@@ -40,23 +192,110 @@ def cos(x: Expr) -> Expr:
...
+    ----------
+    x : relax.Expr
+        The input data
+
+    Returns
+    -------
+    result : relax.Expr
+        The computed result.
+    """
+    return _ffi_api.isnan(x)  # type: ignore
```
Comment (MasterJH5574) [inline python/tvm/relax/op/unary.py]: Let’s move these check functions to the bottom of this file with a separating line of comment :-)

File: src/relax/op/op_common.h [hunk 2]
```diff
@@ -100,12 +99,26 @@ inline TensorStructInfo GetUnaryInputTensorStructInfo(const Call& call, const Bl
...
+      "FInferStructInfo", InferStructInfoUnaryArith<RequireFloatDtype>)
+
+#define RELAX_REGISTER_UNARY_CHECK_OP_AND_IMPL(OpName, RequireFloatDtype) \
+  RELAX_REGISTER_UNARY_OP_AND_IMPL(OpName).set_attr<FInferStructInfo>(    \
+      "FInferStructInfo", InferStructInfoUnaryCheck<RequireFloatDtype>)
+
+#define RELAX_REGISTER_UNARY_NN_OP_AND_IMPL(OpName, OpRegName, RequireFloatDtype) \
+  RELAX_REGISTER_UNARY_OP(OpRegName).set_attr<FInferStructInfo>(                  \
+      "FInferStructInfo", InferStructInfoUnaryArith<RequireFloatDtype>);          \
+  RELAX_UNARY_OP_INTERFACE(OpName, OpRegName);
```
Comment (MasterJH5574) [inline src/relax/op/op_common.h]: After your decoupling, we can now move the two for unary ops to unary.h, and the one for NN to nn.h, as these macros are no longer used across different categories.

File: src/relax/op/op_common.h [hunk 3]
```diff
@@ -100,12 +99,26 @@ inline TensorStructInfo GetUnaryInputTensorStructInfo(const Call& call, const Bl
...
 
-template <bool require_float_dtype>
-inline StructInfo InferStructInfoUnary(const Call& call, const BlockBuilder& ctx) {
+#define RELAX_REGISTER_UNARY_ARITH_OP_AND_IMPL(OpName, RequireFloatDtype) \
+  RELAX_REGISTER_UNARY_OP_AND_IMPL(OpName).set_attr<FInferStructInfo>(    \
+      "FInferStructInfo", InferStructInfoUnaryArith<RequireFloatDtype>)
+
+#define RELAX_REGISTER_UNARY_CHECK_OP_AND_IMPL(OpName, RequireFloatDtype) \
+  RELAX_REGISTER_UNARY_OP_AND_IMPL(OpName).set_attr<FInferStructInfo>(    \
+      "FInferStructInfo", InferStructInfoUnaryCheck<RequireFloatDtype>)
```
Comment (MasterJH5574) [inline src/relax/op/op_common.h]: Given all the check ops don’t require float dtype, we can remove the macro param here.

File: src/relax/op/op_common.h [hunk 4]
```diff
@@ -114,7 +127,30 @@ inline StructInfo InferStructInfoUnary(const Call& call, const BlockBuilder& ctx
...
+        return input_sinfo->dtype;
+      });
+}
+
+template <bool require_float_dtype>
+StructInfo InferStructInfoUnaryCheck(const Call& call, const BlockBuilder& ctx) {
+  return InferStructInfoUnary<require_float_dtype>(
+      call, ctx,
+      [](const Call& call, const BlockBuilder& ctx, const TensorStructInfo& input_sinfo) {
+        return DataType::Bool();
```
Comment (MasterJH5574) [inline src/relax/op/op_common.h]: As you see, we don’t need the call and ctx here.

File: src/relax/op/op_common.h [hunk 5]
```diff
@@ -114,7 +127,30 @@ inline StructInfo InferStructInfoUnary(const Call& call, const BlockBuilder& ctx
         << " requires the input tensor to have float dtype. However, the given input dtype is "
         << input_sinfo->dtype);
   }
-  return input_sinfo;
+  DataType output_dtype = f_compute_out_dtype(call, ctx, input_sinfo);
+  if (input_sinfo->shape.defined()) {
+    return TensorStructInfo(input_sinfo->shape.value(), output_dtype);
+  } else {
+    return TensorStructInfo(output_dtype, input_sinfo->ndim);
+  }
```
Comment (MasterJH5574) [inline src/relax/op/op_common.h]: Your change makes full sense here. Just want to tell you one alternative:
```c++
auto output_sinfo = make_object<TensorStructInfoNode>(*input_sinfo.get());
output_sinfo->dtype = f_compute_out_dtype(input_sinfo);
return TensorStructInfo(output_sinfo);
```

File: src/relax/op/tensor/unary.cc [hunk 6]
```diff
@@ -27,14 +27,37 @@
...
+RELAX_REGISTER_UNARY_ARITH_OP_AND_IMPL(exp, /*require_float_dtype=*/true);
+RELAX_REGISTER_UNARY_ARITH_OP_AND_IMPL(floor, /*require_float_dtype=*/false);
+RELAX_REGISTER_UNARY_ARITH_OP_AND_IMPL(log, /*require_float_dtype=*/true);
+RELAX_REGISTER_UNARY_ARITH_OP_AND_IMPL(negative, /*require_float_dtype=*/false);
+RELAX_REGISTER_UNARY_ARITH_OP_AND_IMPL(round, /*require_float_dtype=*/false);
+RELAX_REGISTER_UNARY_ARITH_OP_AND_IMPL(sigmoid, /*require_float_dtype=*/true);
+RELAX_REGISTER_UNARY_ARITH_OP_AND_IMPL(sign, /*require_float_dtype=*/false);
+RELAX_REGISTER_UNARY_ARITH_OP_AND_IMPL(sin, /*require_float_dtype=*/true);
+RELAX_REGISTER_UNARY_ARITH_OP_AND_IMPL(sinh, /*require_float_dtype=*/true);
+RELAX_REGISTER_UNARY_ARITH_OP_AND_IMPL(square, /*require_float_dtype=*/true);
```
Comment (MasterJH5574) [inline src/relax/op/tensor/unary.cc]: Does square require float dtype 🤔?

**Review by MasterJH5574 — COMMENTED** [review]
Well done, thanks! Just a few quick comments on the macro part.

File: src/relax/op/tensor/unary.cc (hunk 6 above)
Comment (SiriusNEO) [reply src/relax/op/tensor/unary.cc]: It doesn't. Let me change it. At first I think `sqrt` requires so it does as well, but it's incorrect.

File: src/relax/op/op_common.h [hunk 7]
```diff
@@ -127,30 +115,21 @@ inline StructInfo InferStructInfoUnary(const Call& call, const BlockBuilder& ctx
...
 }
 
 template <bool require_float_dtype>
 StructInfo InferStructInfoUnaryCheck(const Call& call, const BlockBuilder& ctx) {
   return InferStructInfoUnary<require_float_dtype>(
-      call, ctx,
-      [](const Call& call, const BlockBuilder& ctx, const TensorStructInfo& input_sinfo) {
-        return DataType::Bool();
-      });
+      call, ctx, [](const TensorStructInfo& input_sinfo) { return DataType::Bool(); });
```
Comment (MasterJH5574) [inline src/relax/op/op_common.h]: Last one: here we can directly pass `/*require_float_dtype=*/false` I think.

Comment (MasterJH5574) [reply src/relax/op/op_common.h]: On line 131.

Comment (SiriusNEO) [reply src/relax/op/op_common.h]: As the template parameter, should we write as `InferStructInfoUnary</*require_float_dtype=*/false>` or directly `InferStructInfoUnary<false>`?

Comment (MasterJH5574) [reply src/relax/op/op_common.h]: false is okay I think 🤦

Comment (SiriusNEO) [reply src/relax/op/op_common.h]: OK since now it is not a template function, I will put its implementation in `op_common.cc`

**Review by MasterJH5574 — APPROVED** [review]

**Comment by SiriusNEO** [comment]
Thanks for reviewing!

**Comment by MasterJH5574** [comment]
Thanks! Would be more than happy if you can help send these changes (except the legalizer part) to tlc-pack/relax.

**Pull request merged**